core/node/da_dispatcher/src/da_dispatcher.rs

```diff
@@ -0,0 +1,199 @@
+use std::{future::Future, time::Duration};
+
+use anyhow::Context;
+use chrono::{NaiveDateTime, Utc};
+use rand::Rng;
+use tokio::sync::watch;
+use zksync_config::DADispatcherConfig;
+use zksync_da_client::{types::DAError, DataAvailabilityClient};
+use zksync_dal::{ConnectionPool, Core, CoreDal};
+use zksync_types::L1BatchNumber;
+
+use crate::metrics::METRICS;
+
+#[derive(Debug)]
+pub struct DataAvailabilityDispatcher {
+    client: Box<dyn DataAvailabilityClient>,
+    pool: ConnectionPool<Core>,
+    config: DADispatcherConfig,
+}
+
+impl DataAvailabilityDispatcher {
+    pub fn new(
+        pool: ConnectionPool<Core>,
+        config: DADispatcherConfig,
+        client: Box<dyn DataAvailabilityClient>,
+    ) -> Self {
+        Self {
+            pool,
+            config,
+            client,
+        }
+    }
+
+    pub async fn run(self, mut stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        let pool = self.pool.clone();
+        loop {
+            if *stop_receiver.borrow() {
+                tracing::info!("Stop signal received, da_dispatcher is shutting down");
+                break;
+            }
+
+            tokio::join!(
+                async {
+                    if let Err(err) = self.dispatch(&pool).await {
+                        tracing::error!("dispatch error {err:?}");
+                    }
+                },
+                async {
+                    if let Err(err) = self.poll_for_inclusion(&pool).await {
+                        tracing::error!("poll_for_inclusion error {err:?}");
+                    }
+                }
+            );
+
+            if tokio::time::timeout(self.config.polling_interval(), stop_receiver.changed())
+                .await
+                .is_ok()
+            {
+                tracing::info!("Stop signal received, da_dispatcher is shutting down");
+                break;
+            }
+        }
+        Ok(())
+    }
+
+    /// Dispatches the blobs to the data availability layer, and saves the blob_id in the database.
+    async fn dispatch(&self, pool: &ConnectionPool<Core>) -> anyhow::Result<()> {
```

Comment:
Why do you need a separate `pool` argument, if it's equal to `self.pool` anyway?